ros/src/computing/perception/localization/packages/kalman_filter_localizer/src/kalman_filter_localizer.cpp

```diff
@@ -0,0 +1,535 @@
+  X_next(IDX::VX) = vx;
+  X_next(IDX::WZ) = wz;
+
+  while (std::fabs(X_next(IDX::YAW)) > M_PI)
+  {
+    X_next(IDX::YAW) -= 2.0 * M_PI * ((X_next(IDX::YAW) > 0) - (X_next(IDX::YAW) < 0));
+  }
+
+  /* Set A matrix for latest state */
+  Eigen::MatrixXd A = Eigen::MatrixXd::Identity(dim_x_, dim_x_);
+  A(IDX::X, IDX::YAW) = -vx * sin(yaw + yaw_bias) * dt;
+  A(IDX::X, IDX::YAWB) = -vx * sin(yaw + yaw_bias) * dt;
+  A(IDX::X, IDX::VX) = cos(yaw + yaw_bias) * dt;
+  A(IDX::Y, IDX::YAW) = vx * cos(yaw + yaw_bias) * dt;
+  A(IDX::Y, IDX::YAWB) = vx * cos(yaw + yaw_bias) * dt;
+  A(IDX::Y, IDX::VX) = sin(yaw + yaw_bias) * dt;
+  A(IDX::YAW, IDX::WZ) = dt;
+
+  /* Set covariance matrix Q for process noise. Calc Q by velocity and yaw angle covariance :
+     dx = Ax + J*w -> Q = J*w_cov*J'          */
+  Eigen::MatrixXd J(2, 2); // coeff of deviation of vx & yaw
+  J << cos(yaw), -vx * sin(yaw),
+      sin(yaw), vx * cos(yaw);
+  Eigen::MatrixXd Q_vx_yaw = Eigen::MatrixXd::Zero(2, 2); // cov of vx and yaw
+  Q_vx_yaw(0, 0) = P_curr(IDX::VX, IDX::VX) * dt * dt;    // covariance of vx
+  Q_vx_yaw(1, 1) = P_curr(IDX::YAW, IDX::YAW) * dt * dt;  // covariance of yaw
+  Eigen::MatrixXd Q = Eigen::MatrixXd::Zero(dim_x_, dim_x_);
+  Q.block(0, 0, 2, 2) = J * Q_vx_yaw * J.transpose(); // for pos_x & pos_y
+  Q(IDX::YAW, IDX::YAW) = cov_proc_yaw_d_;            // for yaw
+  Q(IDX::YAWB, IDX::YAWB) = cov_proc_yaw_bias_d_;     // for yaw bias
+  Q(IDX::VX, IDX::VX) = cov_proc_vx_d_;               // for vx
+  Q(IDX::WZ, IDX::WZ) = cov_proc_wz_d_;               // for wz
+
+  kf_.predictWithDelay(X_next, A, Q);
+}
+
+int KalmanFilterNode::getDelayStep(const double &delay_time, const int &extend_state_step, const double &kf_dt)
+{
+
+  if (delay_time < 0)
+  {
+    ROS_WARN("time stamp is inappropriate, set 0. delay = %f", delay_time);
+    return 0;
+  };
+
+  int delay_step = std::roundf(delay_time / kf_dt);
+  if (delay_step > extend_state_step - 1)
+  {
+    delay_step = extend_state_step - 1;
+    ROS_WARN_DELAYED_THROTTLE(5.0, "delay time: %f[s] exceeds the allowable limit: extend_state_step * kf_dt = %f [s]", delay_time, extend_state_step * kf_dt);
+  }
+
+  return delay_step;
+}
+
+/*
+ * measurementUpdateNDTPose
+ */
+void KalmanFilterNode::measurementUpdateNDTPose(const geometry_msgs::PoseStamped &ndt_pose)
+{
+  // if (ndt_pose.header.frame_id != "world") {
+  //   ROS_WARN_DELAYED_THROTTLE(5, "twist frame_id must be world");
+  // }
+
+  const int dim_y = 3; // pos_x, pos_y, yaw, depending on NDT output
+  const ros::Time t_curr = ros::Time::now();
+
+  /* Calculate delay step */
+  double delay_time = (t_curr - ndt_pose.header.stamp).toSec() + ndt_additional_delay_;
+  int delay_step = getDelayStep(delay_time, extend_state_step_, kf_dt_);
+
+  /* Set yaw */
+  const double yaw_curr = kf_.getXelement((unsigned int)(delay_step * dim_x_ + IDX::YAW));
+  double yaw = tf::getYaw(ndt_pose.pose.orientation);
+  while (std::fabs(yaw - yaw_curr) > M_PI)
```

Comment:
I think that it is better to convert to a vector once and normalize it by atan2
https://github.com/CPFL/Autoware/blob/ae86b06f7638b8b3b4a12381a2844d26ea569f3f/ros/src/common/libs/amathutils_lib/src/vector2d.cpp#L15
https://github.com/CPFL/Autoware/blob/ae86b06f7638b8b3b4a12381a2844d26ea569f3f/ros/src/common/libs/amathutils_lib/src/vector2d.cpp#L30

Reply:
(@sgermanserrano Because there is a lot of such code in autoware, I want to commonize)

Reply:
I changed